src/deepqmc/physics.py

```diff
@@ -31,21 +32,208 @@ def pairwise_self_distance(coords, full=False):
+    vertex) corresponding to N different reference electron.
+    """
+
+    norm = jnp.linalg.norm(rs - nucleus_position, axis=-1)
+    theta = jnp.arccos((rs - nucleus_position)[..., 2] / norm)
+    phi = jnp.arctan2((rs - nucleus_position)[..., 1], (rs - nucleus_position)[..., 0])
+
+    y_rot = jnp.moveaxis(rot_y(theta), -1, -3)  # shape: (3,3,num_e) -> (num_e,3,3)
+
+    z_rot = jnp.moveaxis(rot_z(phi), -1, -3)  # shape: (3,3,num_e) -> (num_e,3,3)
+
+    # auxiliary function applying the rotation and translation to be vmapped
+    def transform_coordinates(norm, z_rot, y_rot, r, nucleus_position):
+        return norm * (z_rot @ y_rot @ r) + nucleus_position
+
+    # vmapping to include N different rotations corresponding to each electron position
+    transform_coordinates = jax.vmap(
+        transform_coordinates, in_axes=(-1, -3, -3, None, None)
+    )
+    # vmapping to be able to transform all 12 icosahedron points at the same time
+    transform_coordinates = jax.vmap(
+        transform_coordinates, in_axes=(None, None, None, -2, None)
+    )
+
+    quadrature_points = transform_coordinates(
+        norm, z_rot, y_rot, UNIT_ICOSAHEDRON, nucleus_position
+    )  # shape: (12,N,3)
+    # we still need to pad the quadrature points with other electron's coordinates
+    quadrature_points_copied = jnp.tile(quadrature_points, (len(rs), 1, 1, 1))
+    rs_copied = jnp.tile(rs, (len(rs), 12, 1, 1))
+    criterion = jnp.moveaxis(
+        jnp.moveaxis(jnp.tile(jnp.eye(len(rs)), (12, 3, 1, 1)), -3, -1), -4, -3
+    )
+    quadrature_rs = jnp.where(
+        criterion, quadrature_points_copied, rs_copied
+    )  # shape: (N,12,N,3)
+
+    return quadrature_rs
+
+
+def nonlocal_potential(rs, mol, state, wf):
+    r"""Calculate the non-local term of the pseudopotential.
+
+    Formulas are based on data from [Burkatzki et al. 2007] or
+    [Annaberdiyev et al. 2018]. Numerical calculation of integrals is based
+    on [Li et al. 2022] where 12-point icosahedron quadrature is used.
+    The current implementation is using jax.lax.fori_loop instead of vmap over
+    the index of the rotated electron. This causes roughly 10% slowdown compared
+    to plain vmap, but avoids OOM issues. Further OOM errors could be resolved by
+    replacing the remaining vmap with fori_loop over the 12 quadrature points.
+
+    Args:
+        rs (float (:math:`N_\text{elec}`, 3)): electron coordinates.
+        mol (:class:`deepqmc.Molecule`): a molecule that is used to load the
+            pseudopotential parameters and nuclear positions.
+        state (dic): wave function state.
+        wf (deepqmc.wf.WaveFunction): the wave function ansatz.
+    """
+
+    # get value of the denominator (which is constant)
+    denominator_wf_sign, denominator_wf_exponent = wf(state, rs)
+
+    # vmap over 12 integration quadrature points
+    wf_vmapped = jax.vmap(wf, in_axes=(None, 0), out_axes=(0))
+
+    total_nl_potential = 0
+    for nucleus_index in range(mol.n_nuc):
```

Comment:
Could it make sense to use a `jax.lax.fori_loop` here to improve the jitting?

Reply:
The refactor from python `for` loop to `lax.fori_loop` is not as straightforward as it might seem. The problem is that different atoms have different numbers of pseudopotential terms. I have worked around this difficulty by using a differently sized array for each atom (I store these arrays in the list `mol.pp_nl_params`) and applying the `.sum()` method to each array. So the `lax.fori_loop` doesn't work here because it can only be applied to pure functions with fixed size arrays. I haven't found any clever workaround yet. I could make `mol.pp_nl_params` an array by padding it with zeros. But that would add a lot of unnecessary terms that have zero contribution. From my not very thorough testing, Jax doesn't optimise the zero terms, but the increase in computational cost from the additional pseudopotential terms is extremely small. So I think it makes sense to rewrite this using `lax.fori_loop` to improve the compile time whilst insignificantly increasing the calculation time. The code would also be in a better agreement with the code for local part of the pseudopotentials, where I used the `lax.fori_loop` with padding (additional zero terms). I'll first properly test the slowdown due to zero nonlocal terms before pushing the update.

Reply:
Okay, after some more testing, I realized that the cost of the evaluating of pseudopotentials is FAR from significant. So it's indeed much better to use `lax.fori_loop`, because it really makes no sense to try to optimize this part at the cost of compile time. I have changed the `for` loop to `lax.fori_loop` in 549cd0eeaccacb727d4ea13fbb309cd887fb3e48